torch_xla/csrc/tensor.cpp

```diff
@@ -840,7 +840,15 @@ bool XLASymNodeImpl::bool_() {
   return dn->getDynamicValue() != 0;
 }
 
-bool XLASymNodeImpl::has_hint() { return true; }
+// "a SymInt has_hint" is equivalent to "a SymInt is backed". Unbacked SymInt is
```

Comment:
From Ed: If the ops are hinted, then the outputs are hinted. Starting point: 1) Inputs are hinted. 2) Data dependent ops are all false. (nonzero is special, 1st dim is unbacked, 2nd dim is backed.)

Reply:
Hi @ezyang , I'm not sure if I follow your suggestion. But before delving into it, it seems torch_xla only creates an `XLASymNodeImpl` when the dimension is dynamic ([code](https://github.com/pytorch/xla/blob/aa532ff18da0c909efcd100de817e1703228fa40/torch_xla/csrc/tensor_impl.cpp#L191-L196)). IOW, as long as we have an `XLASynNodeImpl`, the dimension it represent is dynamic. So would it make sense to you that our `has_hint()` always return false?

Reply:
Uhh, so I guess my question is, does XLA support compilation with varying batch size without recompiling the graph every time? If the answer is no, then yes I agree with your assessment: you can have has_hint always return false. But this is unexpected, and certainly not how dynamo dynamic shapes works. If you are going to do this, it is very very important that dimensions are rarely dynamic (only after, e.g., nonzero call).

Reply:
> does XLA support compilation with varying batch size without recompiling the graph every time?

The answer is no. As I talked with Milad, for the 2 types of dynamism: the one introduced by dynamic ops (such as nonzero) and the one introduced by dynamic input (such as the varying batch size that you mentioned), current torch_xla dynamic shape design only intends to address the former. For latter, we may address it in the future. So I'll try to make `has_hint()` always return false for now.

> If you are going to do this, it is very very important that dimensions are rarely dynamic (only after, e.g., nonzero call).

Thanks for the headsup. Right now, we only have 2 ops `nonzero` and `unique` that introduce dynamic dimensions.

Reply:
In PyTorch proper, we return an opaque IsNonOverlappingAndDense(...) size expression and hope that we don't have to ever actually evaluate it.

Reply:
By making `has_hint()` return false, I discovered that I need to implement the missing methods on XLASymNodeImpl:  IsNonOverlappingAndDense(), wrap_bool(), sym_or(), which may some effort.

So my first question is: is making `has_hint()` return false the correct implementation to you (considering you said `But this is unexpected, and certainly not how dynamo dynamic shapes works.`)? I asked because I don't want to take a seemingly short path and have to revisit this part later. If it's not the correct implementation, can you shed some light on the correct implementation?

Reply:
Yes, I expected that you would have to do this. I guess we should have a call to talk through what is going on with these at a high level.

Reply:
Thanks Ed. Really appreciate it. Maybe we can chat on Tuesday's meeting between PyTorch/XLA and PyTorch? Meanwhile, I can prepare my question in the [doc](https://docs.google.com/document/d/1KQlTojpwDQgauL0RKu1zrq9M2Cf3E7PUkliycV2Clsk/edit).